doc/Parameters.md
## Parameters

Description of the node parameters and how to tune them.

**This document is work-in-progress.**

## Likelihood calculation

- **match\_dist\_min** (float, default: `0.2`, unit: meter):\
    Point to map distance larger than *match\_dist\_min* is ignored in the likelihood calculation.
- **match\_dist\_flat** (float, default: `0.05`, unit: meter):\
    Point-to-map distance smaller than *match\_dist\_flat* is clipped to *match\_dist\_flat*. Set to three-sigma of the input pointcloud noise or larger to avoid fixing to local minimums.

## Expansion resetting

- **match\_ratio\_thresh** (float, default: `0.0`, unit: ratio):\
    Trigger the expansion resetting if the ratio of the measured points which have corresponding map points is smaller than *match\_ratio\_thresh*. Set 0.0 to disable the expansion resetting.
- **expansion\_var\_(x|y|z)** (float, default: `0.2`, unit: meter):\
    Gaussian noise distribution to expand the particle distribution on the expansion resetting.
- **expansion\_var\_(roll|pitch|yaw)** (float, default: `0.05`, unit: radian):\
    Gaussian noise distribution to expand the particle distribution on the expansion resetting.

## Motion prediction

- **odom\_err\_lin\_lin** (float, default: `0.1`, unit: meter/meter):\
    Standard deviation of the translational error per unit of the translational component of the robot's motion.
- **odom\_err\_lin\_ang** (float, default: `0.05`, unit: radian/meter):\
    Standard deviation of the rotational error per unit of the translational component of the robot's motion.
- **odom\_err\_ang\_lin** (float, default: `0.05`, unit: meter/radian):\
    Standard deviation of the translational error per unit of the rotational component of the robot's motion.
- **odom\_err\_ang\_ang** (float, default: `0.05`, unit: radian/rad):\
    Standard deviation of the rotational error per unit of the rotational component of the robot's motion.
- **odom\_err\_integ\_lin\_tc** (float, default: `10.0`, unit: second):\
    Time constant to forget the integral of the translational odometry error.
- **odom\_err\_integ\_lin\_sigma** (float, default: `100.0`, unit: meter):\
    Acceptable range of the integral of the translational odometry error.
- **odom\_err\_integ\_ang\_tc** (float, default: `10.0`, unit: second):\
    Time constant to forget the integral of the rotational odometry error.
- **odom\_err\_integ\_ang\_sigma** (float, default: `100.0`, unit: radian):\
    Acceptable range of the integral of the rotational odometry error.

## ROS communication

- **odom\_queue\_size** (int, default: `200`, unit: count):\
    Odometry subscriber queue size
- **imu\_queue\_size** (int, default: `200`, unit: count):\
    IMU subscriber queue size
- **cloud\_queue\_size** (int, default: `100`, unit: count):\
    Point cloud subscriber queue size


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package mcl_3dl
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.5.4 (2021-03-07)
------------------
* Add odom/imu/cloud_queue_size params (`#375 <https://github.com/at-wat/mcl_3dl/issues/375>`_)
* Fix flaky tests (`#376 <https://github.com/at-wat/mcl_3dl/issues/376>`_)
* Contributors: Atsushi Watanabe

0.5.3 (2021-02-26)
------------------
* Fix cloud accum reset on map update (`#371 <https://github.com/at-wat/mcl_3dl/issues/371>`_)
* Contributors: Atsushi Watanabe

0.5.2 (2021-01-15)
------------------
* Fix potential "Time is out of dual 32-bit range" error (`#367 <https://github.com/at-wat/mcl_3dl/issues/367>`_)
* Update assets to v0.1.4 (`#365 <https://github.com/at-wat/mcl_3dl/issues/365>`_)
* Improve test stability (`#363 <https://github.com/at-wat/mcl_3dl/issues/363>`_)
* Update assets to v0.1.3 (`#362 <https://github.com/at-wat/mcl_3dl/issues/362>`_)
* Update assets to v0.1.2 (`#361 <https://github.com/at-wat/mcl_3dl/issues/361>`_)
* Migrate to GitHub Actions (`#357 <https://github.com/at-wat/mcl_3dl/issues/357>`_)
* Update assets to v0.0.10 (`#356 <https://github.com/at-wat/mcl_3dl/issues/356>`_)
* Contributors: Atsushi Watanabe

0.5.1 (2020-10-26)
------------------
* Make hit_range independent from grid size and fix DDA hit/miss state (`#350 <https://github.com/at-wat/mcl_3dl/issues/350>`_)
* Fix crushing when lidar poses are out of map (`#351 <https://github.com/at-wat/mcl_3dl/issues/351>`_)
* Contributors: Atsushi Watanabe, Naotaka Hatao

0.5.0 (2020-10-15)
------------------
* Fix crushing when new map is received (`#347 <https://github.com/at-wat/mcl_3dl/issues/347>`_)
* Ease condition for test of PointCloudSamplerWithNormal (`#344 <https://github.com/at-wat/mcl_3dl/issues/344>`_)
* Add faster raycast algorithm using DDA (`#343 <https://github.com/at-wat/mcl_3dl/issues/343>`_)
* Contributors: Naotaka Hatao

0.4.0 (2020-10-07)
------------------
* Fix typos of license (`#340 <https://github.com/at-wat/mcl_3dl/issues/340>`_)
* Add PointCloudSamplerWithNormal (`#339 <https://github.com/at-wat/mcl_3dl/issues/339>`_)
* Contributors: Naotaka Hatao

0.3.0 (2020-09-07)
------------------
* Switch beam model by map label field (`#334 <https://github.com/at-wat/mcl_3dl/issues/334>`_)
* Update test script for latest catkin (`#333 <https://github.com/at-wat/mcl_3dl/issues/333>`_)
* Remove references to sensor_msgs::PointCloud (`#332 <https://github.com/at-wat/mcl_3dl/issues/332>`_)
* Update assets to v0.0.9 (`#331 <https://github.com/at-wat/mcl_3dl/issues/331>`_)
* Improve expansion resetting/global localization test stability (`#330 <https://github.com/at-wat/mcl_3dl/issues/330>`_)
* Fix global localization test parameter (`#328 <https://github.com/at-wat/mcl_3dl/issues/328>`_)
* Avoid rate limit when fetching gh-ph-comment (`#329 <https://github.com/at-wat/mcl_3dl/issues/329>`_)
* Update gh-pr-comment (`#327 <https://github.com/at-wat/mcl_3dl/issues/327>`_)
* Retry codecov script download (`#326 <https://github.com/at-wat/mcl_3dl/issues/326>`_)
* Improve test coverage (`#325 <https://github.com/at-wat/mcl_3dl/issues/325>`_)
* Merge rostest coverage profiles (`#324 <https://github.com/at-wat/mcl_3dl/issues/324>`_)
* Contributors: Atsushi Watanabe, f-fl0

0.2.5 (2020-05-27)
------------------
* Add validation for orientation of initial pose (`#317 <https://github.com/at-wat/mcl_3dl/issues/317>`_)
* Update CI scripts (`#318 <https://github.com/at-wat/mcl_3dl/issues/318>`_)
* Contributors: Atsushi Watanabe, Yuta Koga

0.2.4 (2020-05-08)
------------------
* Fix resampling failure of last particle (`#313 <https://github.com/at-wat/mcl_3dl/issues/313>`_)
* Retry gpg keyserver on prerelease test (`#312 <https://github.com/at-wat/mcl_3dl/issues/312>`_)
* Add filter class for Vec3 (`#311 <https://github.com/at-wat/mcl_3dl/issues/311>`_)
* Refactor math functions (`#310 <https://github.com/at-wat/mcl_3dl/issues/310>`_)
* Fix deprecation warning (`#309 <https://github.com/at-wat/mcl_3dl/issues/309>`_)
* Split parameter loader code (`#307 <https://github.com/at-wat/mcl_3dl/issues/307>`_)
* Contributors: Atsushi Watanabe, Naotaka Hatao

0.2.3 (2020-04-07)
------------------
* Update assets to v0.0.8 (`#303 <https://github.com/at-wat/mcl_3dl/issues/303>`_)
* Fix flaky rostest nodes (`#302 <https://github.com/at-wat/mcl_3dl/issues/302>`_)
* Update E2E test parameters (`#301 <https://github.com/at-wat/mcl_3dl/issues/301>`_)
* Refactor CI scripts (`#300 <https://github.com/at-wat/mcl_3dl/issues/300>`_)
* Add Noetic CI job (`#296 <https://github.com/at-wat/mcl_3dl/issues/296>`_)
* Fix initialization of accumulated cloud header (`#299 <https://github.com/at-wat/mcl_3dl/issues/299>`_)
* Support Noetic (`#297 <https://github.com/at-wat/mcl_3dl/issues/297>`_)
* Contributors: Atsushi Watanabe

0.2.2 (2020-03-30)
------------------
* Make average number of accumulated clouds accurate (`#293 <https://github.com/at-wat/mcl_3dl/issues/293>`_)
* Fix latching flag in demo bag (`#294 <https://github.com/at-wat/mcl_3dl/issues/294>`_)
* Fix cloud accumulation logic (`#290 <https://github.com/at-wat/mcl_3dl/issues/290>`_)
* Contributors: Atsushi Watanabe

0.2.1 (2020-02-03)
------------------
* Set DiagnosticStatus::OK as default (`#283 <https://github.com/at-wat/mcl_3dl/issues/283>`_)
* Update assets to v0.0.7 (`#282 <https://github.com/at-wat/mcl_3dl/issues/282>`_)
* Contributors: Atsushi Watanabe, Daiki Maekawa

0.2.0 (2020-01-18)
------------------
* Install consistent version of ros_buildfarm (`#281 <https://github.com/at-wat/mcl_3dl/issues/281>`_)
* Run prerelease test with latest msgs package (`#278 <https://github.com/at-wat/mcl_3dl/issues/278>`_)
* Expose internal errors and convergence status (`#265 <https://github.com/at-wat/mcl_3dl/issues/265>`_)
* Document motion prediction model parameters (`#277 <https://github.com/at-wat/mcl_3dl/issues/277>`_)
* Contributors: Atsushi Watanabe, Daiki Maekawa

0.1.7 (2020-01-06)
------------------

src/parameters.cpp
/*
 * Copyright (c) 2016-2020, the mcl_3dl authors
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of the copyright holder nor the names of its
 *       contributors may be used to endorse or promote products derived from
 *       this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <algorithm>
#include <limits>
#include <memory>
#include <string>

#include <ros/ros.h>

#include <mcl_3dl/parameters.h>

#include <mcl_3dl_compat/compatibility.h>

namespace mcl_3dl
{
bool Parameters::load(ros::NodeHandle& pnh)
{
  pnh.param("fake_imu", fake_imu_, false);
  pnh.param("fake_odom", fake_odom_, false);
  if (fake_imu_ && fake_odom_)
  {
    ROS_ERROR("One of IMU and Odometry must be enabled");
    return false;
  }

  pnh.param("map_frame", frame_ids_["map"], std::string("map"));
  pnh.param("robot_frame", frame_ids_["base_link"], std::string("base_link"));
  pnh.param("odom_frame", frame_ids_["odom"], std::string("odom"));
  pnh.param("floor_frame", frame_ids_["floor"], std::string("floor"));

  mcl_3dl_compat::paramRename<double>(pnh, "likelihood/clip_near", "clip_near");
  mcl_3dl_compat::paramRename<double>(pnh, "likelihood/clip_far", "clip_far");
  mcl_3dl_compat::paramRename<double>(pnh, "likelihood/clip_z_min", "clip_z_min");
  mcl_3dl_compat::paramRename<double>(pnh, "likelihood/clip_z_max", "clip_z_max");
  mcl_3dl_compat::paramRename<double>(pnh, "likelihood/match_dist_min", "match_dist_min");
  mcl_3dl_compat::paramRename<double>(pnh, "likelihood/match_dist_flat", "match_dist_flat");
  mcl_3dl_compat::paramRename<double>(pnh, "likelihood/match_weight", "match_weight");
  mcl_3dl_compat::paramRename<double>(pnh, "likelihood/num_points", "num_points");
  mcl_3dl_compat::paramRename<double>(pnh, "likelihood/num_points_global", "num_points_global");

  mcl_3dl_compat::paramRename<double>(pnh, "beam/clip_near", "clip_beam_near");
  mcl_3dl_compat::paramRename<double>(pnh, "beam/clip_far", "clip_beam_far");
  mcl_3dl_compat::paramRename<double>(pnh, "beam/clip_z_min", "clip_beam_z_min");
  mcl_3dl_compat::paramRename<double>(pnh, "beam/clip_z_max", "clip_beam_z_max");
  mcl_3dl_compat::paramRename<double>(pnh, "beam/num_points", "num_points_beam");
  mcl_3dl_compat::paramRename<double>(pnh, "beam/beam_likelihood", "beam_likelihood");
  mcl_3dl_compat::paramRename<double>(pnh, "beam/ang_total_ref", "ang_total_ref");

  pnh.param("map_downsample_x", map_downsample_x_, 0.1);
  pnh.param("map_downsample_y", map_downsample_y_, 0.1);
  pnh.param("map_downsample_z", map_downsample_z_, 0.1);
  pnh.param("downsample_x", downsample_x_, 0.1);
  pnh.param("downsample_y", downsample_y_, 0.1);
  pnh.param("downsample_z", downsample_z_, 0.05);
  map_grid_min_ =
      std::min(
          std::min(map_downsample_x_, map_downsample_y_),
          map_downsample_z_);
  map_grid_max_ =
      std::max(
          std::max(map_downsample_x_, map_downsample_y_),
          map_downsample_z_);

  pnh.param("update_downsample_x", update_downsample_x_, 0.3);
  pnh.param("update_downsample_y", update_downsample_y_, 0.3);
  pnh.param("update_downsample_z", update_downsample_z_, 0.3);

  double map_update_interval_t;
  pnh.param("map_update_interval_interval", map_update_interval_t, 2.0);
  map_update_interval_.reset(new ros::Duration(map_update_interval_t));

  pnh.param("dist_weight_x", dist_weight_[0], 1.0f);
  pnh.param("dist_weight_y", dist_weight_[1], 1.0f);
  pnh.param("dist_weight_z", dist_weight_[2], 5.0f);
  dist_weight_[3] = 0.0;

  pnh.param("global_localization_grid_lin", global_localization_grid_, 0.3);
  double grid_ang;
  pnh.param("global_localization_grid_ang", grid_ang, 0.524);
  global_localization_div_yaw_ = std::lround(2 * M_PI / grid_ang);

  pnh.param("num_particles", num_particles_, 64);

  pnh.param("resample_var_x", resample_var_x_, 0.05);
  pnh.param("resample_var_y", resample_var_y_, 0.05);
  pnh.param("resample_var_z", resample_var_z_, 0.05);
  pnh.param("resample_var_roll", resample_var_roll_, 0.05);
  pnh.param("resample_var_pitch", resample_var_pitch_, 0.05);
  pnh.param("resample_var_yaw", resample_var_yaw_, 0.05);
  pnh.param("expansion_var_x", expansion_var_x_, 0.2);
  pnh.param("expansion_var_y", expansion_var_y_, 0.2);
  pnh.param("expansion_var_z", expansion_var_z_, 0.2);
  pnh.param("expansion_var_roll", expansion_var_roll_, 0.05);
  pnh.param("expansion_var_pitch", expansion_var_pitch_, 0.05);
  pnh.param("expansion_var_yaw", expansion_var_yaw_, 0.05);
  pnh.param("match_ratio_thresh", match_ratio_thresh_, 0.0);

  pnh.param("odom_err_lin_lin", odom_err_lin_lin_, 0.10);
  pnh.param("odom_err_lin_ang", odom_err_lin_ang_, 0.05);
  pnh.param("odom_err_ang_lin", odom_err_ang_lin_, 0.05);
  pnh.param("odom_err_ang_ang", odom_err_ang_ang_, 0.05);

  pnh.param("odom_err_integ_lin_tc", odom_err_integ_lin_tc_, 10.0);
  pnh.param("odom_err_integ_lin_sigma", odom_err_integ_lin_sigma_, 100.0);
  pnh.param("odom_err_integ_ang_tc", odom_err_integ_ang_tc_, 10.0);
  pnh.param("odom_err_integ_ang_sigma", odom_err_integ_ang_sigma_, 100.0);

  pnh.param("lpf_step", lpf_step_, 16.0);
  pnh.param("acc_lpf_step", acc_lpf_step_, 128.0);

  pnh.param("acc_var", acc_var_, M_PI / 4.0);  // 45 deg

  pnh.param("jump_dist", jump_dist_, 1.0);
  pnh.param("jump_ang", jump_ang_, 1.57);
  pnh.param("fix_dist", fix_dist_, 0.2);
  pnh.param("fix_ang", fix_ang_, 0.1);
  pnh.param("bias_var_dist", bias_var_dist_, 2.0);
  pnh.param("bias_var_ang", bias_var_ang_, 1.57);

  pnh.param("skip_measure", skip_measure_, 1);
  pnh.param("accum_cloud", accum_cloud_, 1);
  pnh.param("total_accum_cloud_max", total_accum_cloud_max_, accum_cloud_ * 10);

  double match_output_interval_t;
  pnh.param("match_output_interval_interval", match_output_interval_t, 0.2);
  match_output_interval_.reset(new ros::Duration(match_output_interval_t));

  double tf_tolerance_t;
  pnh.param("tf_tolerance", tf_tolerance_t, 0.05);
  tf_tolerance_.reset(new ros::Duration(tf_tolerance_t));

  pnh.param("match_output_dist", match_output_dist_, 0.1);
  pnh.param("unmatch_output_dist", unmatch_output_dist_, 0.5);

  pnh.param("publish_tf", publish_tf_, true);
  pnh.param("output_pcd", output_pcd_, false);

  const float float_max = std::numeric_limits<float>::max();
  pnh.param("std_warn_thresh_xy", std_warn_thresh_[0], float_max);
  pnh.param("std_warn_thresh_z", std_warn_thresh_[1], float_max);
  pnh.param("std_warn_thresh_yaw", std_warn_thresh_[2], float_max);

  pnh.param("map_chunk", map_chunk_, 20.0);

  double x, y, z;
  double roll, pitch, yaw;
  double v_x, v_y, v_z;
  double v_roll, v_pitch, v_yaw;
  pnh.param("init_x", x, 0.0);
  pnh.param("init_y", y, 0.0);
  pnh.param("init_z", z, 0.0);
  pnh.param("init_roll", roll, 0.0);
  pnh.param("init_pitch", pitch, 0.0);
  pnh.param("init_yaw", yaw, 0.0);
  pnh.param("init_var_x", v_x, 2.0);
  pnh.param("init_var_y", v_y, 2.0);
  pnh.param("init_var_z", v_z, 0.5);
  pnh.param("init_var_roll", v_roll, 0.1);
  pnh.param("init_var_pitch", v_pitch, 0.1);
  pnh.param("init_var_yaw", v_yaw, 0.5);
  initial_pose_ = State6DOF(
      Vec3(x, y, z),
      Quat(Vec3(roll, pitch, yaw)));
  initial_pose_std_ = State6DOF(
      Vec3(v_x, v_y, v_z),
      Vec3(v_roll, v_pitch, v_yaw));

  pnh.param("use_random_sampler_with_normal", use_random_sampler_with_normal_, false);

  return true;
}
}  // namespace mcl_3dl
